### ISSUE TITLE
Resume training from checkpoint

### ISSUE THREAD

**Issue by michaelklachko** [ppwwyyxx/tensorpack#660]
Currently, when log directory exists, a user is offered a choice to keep/backup/delete/etc. 

If I choose "keep", the message will say:
`Found training history from JSON, now starting from epoch number n.`
(where n in the last epoch number from the previous training session for this model.) 

It makes it seem like we continue the training process from that point, however, in reality we start training from scratch:
https://github.com/ppwwyyxx/tensorpack/blob/efaf9d7ba0dbc5788395ad02a2371853e4ba007a/tensorpack/utils/logger.py#L111

A more useful option would be to actually resume training from the last saved checkpoint.

**Comment by ppwwyyxx** [comment]
That's impossible from a design perspective. All the modules are independent of each other. 

Json loader continues the old epoch number so that new data can be appended to the json file, and this has nothing to do with the rest of the world. Maybe the message can be changed a bit to avoid confusion. And frankly I'm not a big fan of this epoch-loading feature since it has caused other troubles in the past.

**Comment by michaelklachko** [comment]
OK, let's change the message then. Also, the checkpoint names are not very descriptive. It would probably be better to let a user enter a name for a backup/new checkpoint.

**Comment by ppwwyyxx** [comment]
What do you mean by "checkpoint names"? Do you mean the logging directory name?
You can use `set_logger_dir` to set directory name.

**Comment by michaelklachko** [comment]
Yes. For example, right now:
```
if act == 'b':
            backup_name = dirname + _get_time_str()
            shutil.move(dirname, backup_name)
```
After multiple runs with different hyperparams it's hard to determine which backup refers to which run. I have to look into log.log.
It would be better if instead I could simply enter by hand the name of the directory that reflects this particular choice of hyperparams, so that when I want to load a model I can immediately see which checkpoint has it.

Not a big deal though, just a thought.

**Comment by ppwwyyxx** [comment]
You can use `logger.set_logger_dir`, perhaps with a command line argument to do this.

Interacting with users is usually a bad idea. Despite all the other trouble it could cause, it assumes the existence of a user in front of the terminal, which isn't necessarily true at the first place.

**Comment by michaelklachko** [comment]
I agree in general, but in this case you already ask for user input. My suggestion is to ask for the dir name, if the action answer is "backup" or "new". If you don't think that's useful, fine, let's just change the "starting from epoch number n" message and close this ticket.

By the way, I'm curious, what issues did you mean when you said: "I'm not a big fan of this epoch-loading feature since it has caused other troubles in the past."

**Comment by ppwwyyxx** [comment]
Yes in this case it's already interacting with users, which I don't like. I actually have never used "backup" or "new" myself. They turn out to be bad ideas. I usually used `set_logger_dir` and mentally made sure the name doesn't conflict. If I were to rewrite it I would just throw an error when names conflict.

The message was changed to "Rename the first epoch of this training to epoch #{}.".format(epoch)". Hope this won't give users an impression that the checkpoint will be loaded.

The epoch number is by design not supposed to be changed dynamically. The current way of changing it is a bit hacky. And because of this change, callbacks that run before and after the `JSONWriter` will see different epoch numbers which is troublesome. Also this overwrites the `starting_epoch` even if users have manually set it, which is also undesirable.

**Comment by michaelklachko** [comment]
Actually, the option "keep" does not make sense to me. Why is it needed?

**Comment by ppwwyyxx** [comment]
To resume training, if you load the last checkpoint and select "keep", everything will appear as if the training was never interrupted.

**Comment by michaelklachko** [comment]
Oh, I see. This should probably be mentioned somewhere in documentation.

**Comment by ppwwyyxx** [comment]
> everything will appear as if the training was never interrupted.

Not exactly.. because how hacky the epoch number is restored, the new epoch number is not visible to hyperparametersetter. Therefore if the learning rate is scheduled by epoch, it will be wrong after restore.

Just one more reason for me to remove the feature...
